### PR TITLE
VS Code Addon manager

### PR DESCRIPTION
**Pull request by carsakiller**
Adds an addon manager to VS Code that allows users to install addons from [LLS-Addons](https://github.com/LuaLS/LLS-Addons). 

Requires Git to be installed.

### CODE REVIEW

**Comment by carsakiller** [comment]
@sumneko
I can't seem to [get `setConfig` to work](https://github.com/LuaLS/vscode-lua/pull/76/files#diff-0a542fb04a7cc1afc7fa2a0c55236b3199ed14480d8bd55ee506c0af0d14f459R49-R55) with an object.


I printed the value and it looks fine:
```ts
{
  action: "prop",
  key: "Lua.runtime.special",
  prop: "love.filesystem.load",
  value: "loadfile",
  uri: folder.uri
}
```
It returns successfully, but the setting just ends up being an empty array `[]`

**Comment by sumneko** [comment]
I will check it tomorrow

File: client/src/addon_manager/services/settings.service.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+                uri: folder.uri,
+            });
+        } else if (typeof newValue === "object") {
+            changes.push(
+                ...Object.entries(newValue).map(([key, value]) => {
+                    return {
+                        action: "prop",
+                        key: newKey,
+                        prop: key,
+                        value,
```
Comment (sumneko) [inline client/src/addon_manager/services/settings.service.ts:54]: It seems wrong, but why it can be compiled by TS?

**Comment by carsakiller** [comment]
GIF showing usage:
addonManager

**Comment by carsakiller** [comment]
Okay, I think the addon manager is good to go! I think there are just some smaller tasks to make sure everyone knows what is going on. Not sure if you want to wait for these to be done before releasing, up to you. I have written a little announcement post for when the addon manager goes live. It just briefly explains what it is and sort of how it works. This leaves a few things to address:

- [ ] Remove built-in libraries?
  - [ ] Announce removal of built-in libraries
  - [ ] Update wiki to deprecate built-in libraries (I have it nearly complete)
  
- [ ] Add wiki page that explains “Addons”
  - [ ] Explain addon manager
  - [ ] Explain LLS-Addons
  - [ ] Explain LuaCATS organization

**Comment by sumneko** [comment]
Unfortunately, I can't get it work

**Comment by carsakiller** [comment]
🥴 yikes. I built, packaged it, and tested it on a completely fresh VM and things worked smoothly. Does it work at all? Highly possible I screwed up the setting for enabling it (by default)😆.

If you can provide the log file, I can check it out in the morning ☕
image

**Comment by sumneko** [comment]
image

It is a empty page and no output

**Comment by sumneko** [comment]
I found I didn't update the submodule `webvue`.
However I don't know how to compile it.

**Comment by carsakiller** [comment]
Ah, are you not using the packaged `vsix`?

**Comment by sumneko** [comment]
> Ah, are you not using the packaged `vsix`?

No, I'm using `installExtensionFromLocation`

**Comment by carsakiller** [comment]
Hm, that should work then. The actions should have compiled the TS and built webvue.

You will want to run `tsc` in `client/`. Then `cd` to `client/webvue` and run `npm run build`. Then you'll have to repackage `vscode-lua`.

Alternatively you can run it in debug mode but it doesn't look accurate (especially the colours):
1. `tsc --watch` in `client/`
2. `npm run dev` in `client/webvue`
3. <kbd>F5</kbd> to open extension debug host

**Comment by sumneko** [comment]
OK, it works!

**Comment by sumneko** [comment]
Because I runed `tsc` instead of `npm run build` before...

**Comment by carsakiller** [comment]
But you installed the `vsix` that you downloaded from the GH action, right? It should be compiled and built already then

**Comment by sumneko** [comment]
> But you installed the `vsix` that you downloaded from the GH action, right? It should be compiled and built already then

No, I always use the project it self.
Command `installExtensionFromLocation` can install `vscode-lua` from your local directory.

**Comment by carsakiller** [comment]
Ahhh, okay. I was confused because I thought you were using "Install from VSIX" to install the GH action artifact 🥴. A bit more detail can be found in the [webvue repo](https://github.com/LuaLS/vscode-lua-webvue#readme). Glad it is working then!

**Comment by sumneko** [comment]
Maybe we need a script to compile client, including `update submodules`, `install npm`, `tsc` and `run build`.

**Comment by carsakiller** [comment]
We could use `npm` so we could say `npm run all` or something and it can run `git submodule update && npm i && tsc && npm run build`

**Comment by carsakiller** [comment]
@sumneko, what will be the new version number that has the addon manager and removes the built-in libraries? I am writing an announcement for the removal of the built-in libraries, but need to know which version it will be included in. I can open a PR that deletes the built-ins if you would like.

**Comment by carsakiller** [comment]
Also, I have somehow managed to activate discussions in [vscode-lua](https://github.com/LuaLS/vscode-lua), despite not having permission to disable it.... I think we should keep discussions to just [lua-language-server](https://github.com/LuaLS/lua-language-server), but issues for the addon manager can live in [vscode-lua](https://github.com/LuaLS/vscode-lua)

**Comment by carsakiller** [comment]
Hey @sumneko!

Here is what I have for the announcement for the removal of the built-in libraries (to be posted ASAP):

<details>
<summary>Announcement</summary>

Hello 👋

With the VS Code addon manager releasing very soon, we have decided to remove the built-in libraries that ship with the language server in the same release.

**This means the `${3rd}` shortcut for [`workspace.library`](https://github.com/LuaLS/lua-language-server/wiki/Settings#workspacelibrary) will no longer work**.

> **Note**
> Older versions of the server that still contain the built-in libraries will continue to function as-is.

# Why?
The addon manager for VS Code is releasing at the same time as this change, meaning everyone using the VS Code extension will still be able to install definition files super easily.

Removing the built-in libraries helps prevent duplicates for those using VS Code and can help cut down on bloat for everyone. It is unlikely that everyone is interested in all the built-in libraries, so many of them end up not being used.

# What is the New Method?
Libraries are now addons. The name “libraries” is a little misleading, as they are actually just definitions (and maybe a plugin).

Those using VS Code can use the addon manager to install addons that contain all the definition files that were included in the language server, and more!

Those using other clients can still download all the libraries that were included in the server from the [LuaCATS organization](https://github.com/orgs/LuaCATS/repositories). They can then be [dropped into a folder](https://github.com/LuaLS/lua-language-server/wiki/Libraries#setup) and linked to with [`workspace.userThirdParty`](https://github.com/LuaLS/lua-language-server/wiki/Settings#workspaceuserthirdparty) for functionality similar to how the built-ins have previously worked, or [manually linked](https://github.com/LuaLS/lua-language-server/wiki/Libraries#link-to-workspace) using [`workspace.library`](https://github.com/LuaLS/lua-language-server/wiki/Settings#workspacelibrary).

# Contribute
Collaboration on definition files is now better than ever.

Do you have definition files for a library? Now is the time to contribute them! If you would like it to show up in the addon manager, you can add it to [LLS-Addons](https://github.com/LuaLS/LLS-Addons#readme).

Do you have definition files that you can no longer maintain? If you transfer ownership to me (@carsakiller), I can transfer the project to [LuaCATS](https://github.com/LuaCATS), where the community can work together to keep the definitions up-to-date.

</details>

---

And for the addon manager (to be posted at release of VS Code extension):

<details>

<summary>Announcement</summary>


New in version X.XX of the VS Code extension is the addon manager! 🎉

It can be opened using <kbd>Ctrl + P</kbd> to open the command palette and searching for the `Open Addon Manager` command.

# Summary

The addon manager allows you to enable, disable, and uninstall addons for the language server. This means you can now easily install definitions for libraries and automatically apply settings to better emulate an environment.

> **Warning**
> While each addon added to [LLS-Addons](https://github.com/LuaLS/LLS-Addons) will be reviewed, it is possible for an addon to modify *any* VS Code setting using its `config.json`. This offers great freedom but is also potentially abusable. Pressing the GitHub button will open the addon, allowing you to review it yourself if you'd like to double-check 🕵️‍♂️

Addon manager usage example

# Contributing Addons
The addon manager pulls from the [LLS-Addons](https://github.com/LuaLS/LLS-Addons) repository and currently contains many of the definitions from the language server. If you would like to contribute an addon, head over to the [LLS-Addons README to learn more](https://github.com/LuaLS/LLS-Addons#readme). They are very similar to the [environment emulations](https://github.com/LuaLS/lua-language-server/wiki/Libraries#environment-emulation) that have been in use for a while, but add a little more configuration and are missing [plugin support](https://github.com/LuaLS/vscode-lua-webvue/issues/5).

# WebVue UI
The addon manager is a [webview](https://code.visualstudio.com/api/extension-guides/webview) (basically an iframe) that allows us to display a custom webpage within VS Code. You can check out the code for the webview UI in the [vscode-lua-webvue repository](https://github.com/LuaLS/vscode-lua-webvue). It uses Vue.js and some VS Code components and styling to try to make it feel as at-home as possible.

# Feedback
The addon manager has been [a while in the making](https://github.com/LuaLS/lua-language-server/discussions/1607), so I'm excited to finally merge it in. If you have any feedback, please leave it in the [VS Code extension repository](https://github.com/LuaLS/vscode-lua) ❤️.

</details>

I also have some wiki changes ready to go that deprecate the built-in libraries and explain addons (which replace the “environment emulations”)

If you know the version that these changes will take place in, I can add it to the announcements.

**Comment by sumneko** [comment]
I can't guarantee the release cycle of the version, so I think it can be declared that it is removed in a medium version update after half a year.

**Comment by carsakiller** [comment]
Okay, makes sense. Do you know which version the addon manager would be released in?

**Comment by sumneko** [comment]
It is not a break change for user, so I can publish a small version for it

**Pull request merged**